### PR TITLE
Add Quill editor to enable rich text content

### PR DESCRIPTION
**Pull request by Neriderc**
This PR adds Quill to replace the textarea for messages (both top level and replies).

Quill is highly extensible but this is only implementing the most basic functions, it's based on the React example in the QuillJS playground [here](https://quilljs.com/playground/react).

Quill is used to display the formatted data. It is loaded conditionally depending on if the message contains rich text, however, _all_ text content is now saved in JSON format with both Quill's "Delta" format and the textOnly equivalent. A fallback exists to handle messages written before Quill was added, however, editing a message will update the message's `text` field to the new format.

This change introduces the message type "rich_text" to identify if Quill needs to be loaded to display the content. "standard" messages are still displayed without Quill, even if stored in JSON format.

I think I have preserved functions tied to the old text area: message previews, paste image from clipboard, ctrl+enter to submit the message.

There isn't currently an issue to link to this, not sure if one should be created to help with tracking this?

Edit:

Adding that this PR also contains some other updates so TL;DR, this PR contains:
- New Quill editor for rich text functionality
- Updated docker install instructions
- Fixed issue where the little pin icon was showing in the feed even though it has no effect there

Please let me know if anything should be spun off to it's own PR

### CODE REVIEW

**Comment by nrobinaubertin** [comment]
> This isn't currently an issue to link to this, not sure if one should be created to help with tracking this?

I have no problem having just the PR

File: app/src/message/message-body.component.js [hunk 1]
```diff
@@ -1,18 +1,20 @@
...
+  const displayStandardMessageText = (text) => {
+    if (!text) {
       return "";
     }
     // escape html a little (just enough to avoid xss I hope)
-    let txt = props.message.data["text"]
+    let txt = text
       .replace(/</g, "&lt;")
       .replace(/>/g, "&gt;")
       .trim();
```
Comment (nrobinaubertin) [inline app/src/message/message-body.component.js]: I don't think that this part is needed anymore. Quill will provide better protection

File: app/src/message/message-preview.component.js [hunk 2]
```diff
@@ -54,6 +55,15 @@ export default function MessagePreview(props) {
       setLoaded(true);
       setLastActivityDate(p?.lastActivityDate);
       setData(p?.data);
+
+      // We need to handle messages created before the editor 
+      // was added and we started storing in JSON
+      try {
+        setTextPreview(JSON.parse(p?.data.text).textOnly);
+      } catch {
+        setTextPreview(p?.data?.text);
```
Comment (nrobinaubertin) [inline app/src/message/message-preview.component.js]: This logic to get the text content of a message is repeated a lot. I think that it would make sense to create a function for it

**Review by nrobinaubertin — COMMENTED** [review]
Some comments just looking at the code but I need to run it myself and I'll do a second review next week

**Comment by Neriderc** [comment]
I've put a few more commits on this since we last talked, but as far as I know this should be all ready to go. I've deployed to my prod server to test it and found no issues so far.

File: app/src/quill/QuillEditor.component.js
Comment (nrobinaubertin) [inline app/src/quill/QuillEditor.component.js]: I would prefer this to be named using snake-case (`quill-editor.component.js`)

**Review by nrobinaubertin — COMMENTED** [review]
I tested it locally and have some comments but your PR seems solid!

Running it on a thread with more than 1k comments and using firefox's profiler, I can see that the memory usage increases quite a bit with quill (380Mo vs 90Mo in my non scientific test). But I would argue that it remains an edge case to have that many comments loaded at once. So I think that it is fine.

File: app/src/quill/quill-read-only.component.js
[the remark this answers is not in the document]
Comment (Neriderc) [reply app/src/quill/quill-read-only.component.js]: Updated

**Comment by nrobinaubertin** [comment]
> Were they just plain text comments?

No, it's a music thread with lots of embeds from youtube. But I think that it's fine. My test was not really a good measurement either way since it is not reproducible

File: app/src/quill/quill-read-only.component.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { h } from "preact";
+import { useEffect, useRef } from "preact/hooks";
+import Quill from "quill";
+import "quill/dist/quill.snow.css";
```
[the remark this answers is not in the document]
Comment (Neriderc) [reply app/src/quill/quill-read-only.component.js]: Done 🙂

**Review by nrobinaubertin — APPROVED** [review]
